manifests/osfamily/windows.pp

```diff
@@ -0,0 +1,37 @@
+class puppet_agent::osfamily::windows {
+
+  $_arch = $::kernelmajversion ?{
+    /^5\.\d+/ => 'x86', # x64 is never allowed on windows 2003
+    default   => $::puppet_agent::arch,
+  }
+
+  $_source = $::puppet_agent::source ? {
+    undef   => "https://downloads.puppetlabs.com/windows/puppet-agent-${_arch}-latest.msi",
+    default => $::puppet_agent::source,
+  }
+
+  $_msi_location = $_source ? {
+    /^puppet:/ => "${tmpdir}\\puppet-agent.msi",
+    default    => $_source,
+  }
+
+  if $_source =~ /^puppet:/ {
+    file{ $_msi_location:
+      source => $_source,
+      before => File["${tmpdir}\\install_puppet.bat"],
+    }
+  }
+
+  $_timestamp = strftime('%Y_%m_%d-%H_%M')
+  $_logfile = "${tmpdir}\\puppet-${_timestamp}-installer.log"
+  debug ("Puppet upgrade log file at ${_logfile}")
+  debug ("Installing puppet from ${_msi_location}")
+  file { "${tmpdir}\\install_puppet.bat":
+    ensure  => file,
+    content => template('puppet_agent/install_puppet.bat.erb')
```

Comment:
I'm hoping that we won't have any issues with line endings. https://tickets.puppetlabs.com/browse/PUP-927 - should be fixed in 3.8.x though.

Otherwise we could have gone the file route and passed it the pid as an argument.